### PR TITLE
Simplify beam search and improve efficiency

### PR DESCRIPTION
**Pull request by bpopeters**
This PR makes several changes to the translation code aimed toward improving readability and maintainability. This includes updating pre-0.4.0 code by removing unnecessary use of `.data` and taking advantage of new tensor creation ops where possible, and also more general things -- many spurious attributes are gone (although I'm sure one who wanted to have a look could find many more). The interface for the `Beam` should also be a little bit clearer. I removed the `PenaltyBuilder` class, as it was essentially a dictionary. Many of the methods of `GNMTGlobalScorer` have been moved to `Beam` in order to avoid circumlocutions like `global_scores = self.global_scorer.score(self, self.scores)`.

I don't currently have the hardware to test the performance rigorously, but I can translate my toy g2p set significantly faster with this code than with the master code with a beam size >= 15.

### CODE REVIEW

**Comment by guillaumekln** [comment]
Hello. For efficiency, there is also a new beam search path behind the `-fast` decoding flag. Did you try it out?

**Comment by bpopeters** [comment]
Your fast beam search is faster than either, which is not surprising. It also records better average scores on both pred and gold, which I wasn't expecting.

Any performance improvements coming from my changes are incidental -- my main goal was to make the translation code clearer. There's still more to be done on that front, of course.

**Comment by guillaumekln** [comment]
> my main goal was to make the translation code clearer.

That's always a good change. I was mentionning the new code path because as far as I know, the goal would be to port all existing options to it and remove the old code path which uses the `Beam` class.

**Comment by bpopeters** [comment]
If so, then some of my alterations will end up being irrelevant. But `Beam` wasn't the only thing I worked on.

**Comment by bpopeters** [comment]
Update: the most recent commit fixes an error in translator.py. Previously, using `-dump_beam` would reference a nonexistent attribute and fail. I fixed this.

**Comment by vince62s** [comment]
When translating on gpu I am getting this:
  File "/home/moses/pytorchwork/Ben/OpenNMT-py/onmt/translate/translator.py", line 588, in _translate_batch
    b.advance(out[:, j], beam_attn[:, j, :memory_lengths[j]])
  File "/home/moses/pytorchwork/Ben/OpenNMT-py/onmt/translate/beam.py", line 116, in advance
    ix = indices.masked_select(finished_sents)
RuntimeError: Expected object of type torch.ByteTensor but found type torch.cuda.ByteTensor for argument #2 'mask'

**Comment by bpopeters** [comment]
This should fix it unless I also forgot to pass device when I needed to elsewhere.

**Comment by vince62s** [comment]
First is your code, second master repo.
Same scores, your code slightly slower (or equivalent)
Based on beam size of 5 and batchsize of 32
Anyway, your code looks cleaner.

PRED AVG SCORE: -0.4147, PRED PPL: 1.5139
BLEU = 35.28, 65.8/42.9/30.6/22.5 (BP=0.945, ratio=0.946, hyp_len=31164, ref_len=32931)
PRED AVG SCORE: -0.3771, PRED PPL: 1.4581
BLEU = 43.06, 69.4/48.2/36.2/28.4 (BP=1.000, ratio=1.000, hyp_len=6057, ref_len=6056)
PRED AVG SCORE: -0.3517, PRED PPL: 1.4215
BLEU = 50.87, 80.5/60.8/48.8/39.5 (BP=0.918, ratio=0.921, hyp_len=1494, ref_len=1622)
PRED AVG SCORE: -0.3388, PRED PPL: 1.4033
BLEU = 49.18, 74.0/55.6/44.3/36.1 (BP=0.971, ratio=0.972, hyp_len=4369, ref_len=4496)
PRED AVG SCORE: -0.3743, PRED PPL: 1.4540
BLEU = 39.33, 68.9/46.8/34.5/26.0 (BP=0.954, ratio=0.955, hyp_len=10353, ref_len=10844)

real	6m0.916s
user	4m54.020s
sys	1m6.116s

PRED AVG SCORE: -0.4147, PRED PPL: 1.5139
BLEU = 35.28, 65.8/42.9/30.6/22.5 (BP=0.945, ratio=0.946, hyp_len=31164, ref_len=32931)
PRED AVG SCORE: -0.3771, PRED PPL: 1.4581
BLEU = 43.06, 69.4/48.2/36.2/28.4 (BP=1.000, ratio=1.000, hyp_len=6057, ref_len=6056)
PRED AVG SCORE: -0.3517, PRED PPL: 1.4215
BLEU = 50.87, 80.5/60.8/48.8/39.5 (BP=0.918, ratio=0.921, hyp_len=1494, ref_len=1622)
PRED AVG SCORE: -0.3388, PRED PPL: 1.4033
BLEU = 49.18, 74.0/55.6/44.3/36.1 (BP=0.971, ratio=0.972, hyp_len=4369, ref_len=4496)
PRED AVG SCORE: -0.3743, PRED PPL: 1.4540
BLEU = 39.33, 68.9/46.8/34.5/26.0 (BP=0.954, ratio=0.955, hyp_len=10353, ref_len=10844)

real	5m50.655s
user	4m43.884s
sys	1m3.896s

**Comment by bpopeters** [comment]
What about with a wider beam?

**Comment by vince62s** [comment]
Beam 10. Yours, master. Very similar. BTW you'll see score is lower than Beam 5....

PRED AVG SCORE: -0.4118, PRED PPL: 1.5095
BLEU = 35.15, 66.0/43.2/30.8/22.7 (BP=0.935, ratio=0.937, hyp_len=30870, ref_len=32931)

real	6m5.046s
user	4m44.076s
sys	1m19.372s

PRED AVG SCORE: -0.4118, PRED PPL: 1.5095
BLEU = 35.15, 66.0/43.2/30.8/22.7 (BP=0.935, ratio=0.937, hyp_len=30870, ref_len=32931)

real	6m10.932s
user	4m50.880s
sys	1m18.696s

**Comment by vince62s** [comment]
Ben,

if you have time, can you please read this. #777
I think we can go even a step further.

In the same range of idea, we could add a parameter to the Translator class, eg translate_loss that we need to initialize to onmt.utils.loss.build_loss_compute(model, fields["tgt"].vocab, opt, train=False)
(same as valid_loss in build_trainer)

Then we woudl have a loss_compute object (that could handle copy_attn cases) and we may remove some duplicated code thats is in _translate_batch() and in _run_target()

Hope this is clear.

**Comment by bpopeters** [comment]
I can't make any promises about this week, but I'll try to take a look at this when I can. Perhaps you could reopen the issue so it's easier for me to remember?

**Comment by vince62s** [comment]
Hi @bpopeters do you have time to finalize this ? many thanks.

**Comment by vince62s** [comment]
@guillaumekln @bpopeters I am going to merge this before it becomes obsolete.
I hope I did not forget any thing in resolving conflicts.
I'll work on #777 as soon as I can.

File: onmt/translate/penalties.py [hunk 1]
```diff
@@ -1,80 +1,57 @@
...
+    return 0. if beta == 0. else beta * -cov.clamp(max=1.0).log().sum(1)
+
+
+def coverage_summary(scores, cov, beta=0.):
+    """
+    Our summary penalty.
     """
-    Returns the Length and Coverage Penalty function for Beam Search.
+    return 0. if beta == 0. else beta * \
+        (-cov.clamp(min=1.0).sum(1) - cov.size(1))
```
Comment (vince62s) [inline onmt/translate/penalties.py]: This triggers an error, clamp does not exist.

File: onmt/translate/penalties.py [hunk 2]
```diff
@@ -1,80 +1,57 @@
...
+    return lengths.get(name, length_none)
+
+
+def coverage_wu(scores, cov, beta=0.):
+    """
+    NMT coverage re-ranking score from
+    "Google's Neural Machine Translation System" :cite:`wu2016google`.
+    """
+    return 0. if beta == 0. else beta * -cov.clamp(max=1.0).log().sum(1)
+
```
Comment (vince62s) [inline onmt/translate/penalties.py]: This triggers an error, clamp does not exist.

File: onmt/translate/beam.py [hunk 3]
```diff
@@ -172,12 +172,57 @@ def get_hyp(self, timestep, k):
...
             hyp.append(self.next_ys[j + 1][k])
             attn.append(self.attn[j][k])
             k = self.prev_ks[j][k]
         return hyp[::-1], torch.stack(attn[::-1])
 
+    def update_score(self, attn):
+        if "prev_penalty" in self.global_state:
+            self.scores.add_(self.global_state["prev_penalty"])
+            penalty = self.global_scorer.cov_penalty(
+                self, self.global_state["coverage"] + attn,
```
Comment (vince62s) [inline onmt/translate/beam.py]: first self should be removed

File: onmt/translate/penalties.py (hunk 1 above)
Comment (bpopeters) [reply onmt/translate/penalties.py]: [Clamp](https://pytorch.org/docs/stable/tensors.html#torch.Tensor.clamp) [exists](https://pytorch.org/docs/stable/torch.html#torch.clamp). What is the value of `cov`?

Comment (vince62s) [reply onmt/translate/penalties.py]: 5.0 it's a flaot in the opts.py

Comment (bpopeters) [reply onmt/translate/penalties.py]: Looking at the history of the penalties code, that must be a recent change.

Comment (sebastianGehrmann) [reply onmt/translate/penalties.py]: I think there is a confusion here - `beta` is the hyperparameter that is a float. `cov` is the sum of attentions from all previous decoding steps. 
The summary penalty is from my recent EMNLP paper and penalizes repeated attention, wu penalty is from the google nmt paper and penalizes not attenting to words. 
@vince62s What is the exact error you are getting?

Comment (vince62s) [reply onmt/translate/penalties.py]: my confusion.
the issue is not here but in beam.py below.

File: onmt/translate/beam.py (hunk 3 above)
Comment (vince62s) [reply onmt/translate/beam.py]: okay I think I know why.
now cov_penalty() takes beam.scores and cov as arguments, not the beal itself, nor the beta
Testing it and will fix this.

Comment (vince62s) [reply onmt/translate/beam.py]: well I chnaged to:
            penalty = self.global_scorer.cov_penalty(
                self.scores, self.global_state["coverage"] + attn)
But even if it is "running" apparently fine, results are not good.

**Comment by Henry-E** [comment]
I would say that for research purposes and modifying the beam search easily this is definitely not irrelevant!

**Comment by vince62s** [comment]
I am very willing to merge this, but 1) there new conflicts to resolve 2) when I tested it with a summarization existing model, scores before / after were not the same.
I do'nt have too much time to dig further. If one can find the issue, that would be great.

**Comment by bpopeters** [comment]
It would be great if this code could get merged, but I wrote this nearly four months ago and refamiliarizing myself with it so I can fix merge errors caused by everything else that changed in the codebase in the meantime is not my highest priority at the present moment.

As for differences with scores, this is something a colleague and I are going to look into in general because he's reporting worse bleu scores when he uses `-fast` translation and I had a problem with duplicate hypotheses in the beam a while ago that I haven't been able to look into thoroughly yet.

**Comment by vince62s** [comment]
When I am doing some refactoring as I did for translator (and I also checked when @guillaumekln did the fast):
I am scoring a DE-EN with fast and non-fast beam search
I am scoring a CNN-DM sumarization with fast and non-fast (with copy_attn, ...)
and checking I get the same results before / after.
I did it also for Gold scores after fixing them.
I am not always checking Ensemble scores.
I dont know how to best do it, but having some automated regression tests may help for all of this.

Anyway, this PR does not give the same results for sumarization for now.

**Pull request closed without merging**